### PR TITLE
feat: introduce typelist

### PR DESCRIPTION
**Pull request by asalzburger**
This PR introduces a minimal implementation of a C++ typelist to be used many in grid I/O (Json/Svg) in order to allow casting to a dedicated grid type.

It follows relatively closely to what @niermann999 has done for detray, but omits the `at[index]` access as it aims to allow type iteration and not individual type casting.

### CODE REVIEW

**Review by paulgessinger — COMMENTED** [review]
This looks pretty good. 

- can you change the detail namespace name something more unique, since the functions themselves have fairly generic names.
- equivalently for the typedefs, they have very generic names like `front`, which we might want to avoid having in the global detail namespace. In the EDM work I went with `detail_xyz` to scope the types. 
- we could consider even putting the `TypeList` type and manipulator functions in the main `Acts` namespace. 

I think this can also be done in a follow-up PR.

File: Tests/UnitTests/Core/Utilities/TypeListTests.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <boost/test/data/test_case.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include "Acts/Utilities/detail/TypeList.hpp"
+
+#include <type_traits>
```
Comment (niermann999) [inline Tests/UnitTests/Core/Utilities/TypeListTests.cpp:14]:
```suggestion
#include <typeinfo>
#include <type_traits>
```

File: Tests/UnitTests/Core/Utilities/TypeListTests.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  class C {};
+
+  using AbList = Acts::detail::TypeList<A, B>;
+  auto abc = Acts::detail::push_back<AbList, C>{};
+  bool backIsC = std::is_same_v<Acts::detail::back<decltype(abc)>, C>;
+  BOOST_CHECK(backIsC);
+}
+
+template <typename Head, typename... Tail>
+void printTypes(
```
Comment (niermann999) [inline Tests/UnitTests/Core/Utilities/TypeListTests.cpp:58]: This output will be a bit cryptic and become unreadable for more complicated types, I guess.

Comment (paulgessinger) [reply Tests/UnitTests/Core/Utilities/TypeListTests.cpp:58]: Right now this is only in the test so it's probably not a big deal. If we promote this to the core lib we'd probably want to demangle with boost.

File: Core/include/Acts/Utilities/detail/TypeList.hpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <tuple>
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Utilities/detail/TypeList.hpp:11]: We can!

File: Tests/UnitTests/Core/Utilities/TypeListTests.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  using MyList = Acts::detail::TypeList<float, int, double, A>;
+
+  bool frontIsFloat = std::is_same_v<Acts::detail::front<MyList>, float>;
+  BOOST_CHECK(frontIsFloat);
+
+  bool backIsA = std::is_same_v<Acts::detail::back<MyList>, A>;
+  BOOST_CHECK(backIsA);
+
+  BOOST_CHECK(Acts::detail::size<MyList> == 4);
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Utilities/TypeListTests.cpp:32]:
```suggestion
  BOOST_CHECK_EQUAL(Acts::detail::size<MyList>, 4);
```

File: Tests/UnitTests/Core/Utilities/TypeListTests.cpp (hunk 2 above)
Comment (asalzburger) [reply Tests/UnitTests/Core/Utilities/TypeListTests.cpp:58]: Sure, this is really only a unit test.

File: Core/include/Acts/Utilities/detail/TypeList.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Utilities/detail/TypeList.hpp]: Done.

**Review by niermann999 — DISMISSED** [review]

**Comment by asalzburger** [comment]
Hi, I should have addressed the first set of comments.

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [6352ad85f4ac8d05f698716afc882227656524e1]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6480951)!

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951696) |
|:red_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951695) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951694) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951693) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951692) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951691) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951690) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951689) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951688) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951687) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951686) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951685) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951684) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951683) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951682) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951681) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951680) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33951679) |

**Pull request merged**